data.go

```diff
@@ -174,49 +183,70 @@ func (d *Data) Datasource(alias string, args ...string) map[string]interface{} {
 	if err != nil {
 		log.Fatalf("Couldn't read datasource '%s': %s", alias, err)
 	}
-	if source.Type == "application/json" {
-		ty := &TypeConv{}
-		return ty.JSON(string(b))
-	}
-	if source.Type == "application/yaml" {
-		ty := &TypeConv{}
-		return ty.YAML(string(b))
-	}
-	log.Fatalf("Datasources of type %s not yet supported", source.Type)
-	return nil
+	return b
 }
 
 // ReadSource -
-func (d *Data) ReadSource(fs vfs.Filesystem, source *Source, args ...string) ([]byte, error) {
-	if d.cache == nil {
-		d.cache = make(map[string][]byte)
-	}
-	cacheKey := source.Alias
-	for _, v := range args {
-		cacheKey += v
-	}
-	cached, ok := d.cache[cacheKey]
-	if ok {
+func (d *Data) ReadSource(fs vfs.Filesystem, source *Source, args ...string) (map[string]interface{}, error) {
```

Comment:
> Also, I think it makes sense to include the (potentially costly) deserialized objects into the cache so that this conversion needs to be called only once (and not for every call to the datasource). I don't consider this optimisation as premature since the call to the datasource can happen quite often also when doing only a single call.

If you're using the same datasource multiple times in a template, it makes sense to assign its output to a variable:

```go
{{ $config := datasource "config" }}
hostname: {{ $config.hostname }}
port: {{ $config.port }}
...
```

So datasources should almost never be called multiple times in a template. It's non-obvious if someone is new to gomplate and go templates in general, but maybe that could be mitigated with documentation for now... Ultimately it would be nice to have some sort of subcommand that could lint the template, but that's a long way off 😉